File: OCP/C#/Example 1/improved.cs
public interface IServer
{
	void ProcessMessage(string message)
}

public class Server: IServer
{
	public void ProcessMessage(string message)
	{
		Console.WriteLine($"Server received: {message}");
	}
}

public class AdvancedServer: IServer
{
	public void ProcessMessage(string message)
	{
		Console.WriteLine($"Server received: {message}");
		Console.WriteLine("Performing advanced server process");
		// Advanced processing...
	}
}

public class Client
{
	private IServer _server
	
	// Dependency injection via constructor
	public Client(IServer server)
	{
		_server = server;
	}
	public void SendMessage(string message)
	{
		_server.ProcessMessage(message)
	}
}

File: OCP/Javascript/Example 1/improved.js
class IServer {
    processMessage(message) {
        throw new Error("Method 'processMessage()' must be implemented.");
    }
}

class Server extends IServer {
    processMessage(message) {
        console.log(`Server received: ${message}`);
    }
}

class AdvancedServer extends IServer {
    processMessage(message) {
        console.log(`Server received: ${message}`);
        console.log("Performing advanced server process");
        // Advanced processing...
    }
}

class Client {
    constructor(server) {
        if (!(server instanceof IServer)) {
            throw new Error("server must be an instance of IServer");
        }
        this._server = server;
    }

    sendMessage(message) {
        this._server.processMessage(message);
    }
}
